File: src/readers/shapefile/shp.ts
// Implements https://www.esri.com/content/dam/esrisites/sitecore-archive/Files/Pdfs/library/whitepapers/pdfs/shapefile.pdf
import { extendBBox } from '../../geometry/index.js';
import { toReader } from '../index.js';

import type { DataBaseFile } from './dbf.js';
import type { Transformer } from '../../proj4/index.js';
import type {
  BBOX,
  BBox3D,
  MValue,
  Properties,
  VectorFeature,
  VectorFeatureCollection,
  VectorGeometry,
  VectorLineString,
  VectorLineStringGeometry,
  VectorMultiLineString,
  VectorMultiLineStringGeometry,
  VectorMultiPoint,
  VectorMultiPointGeometry,
  VectorPoint,
  VectorPointGeometry,
  VectorPolygonGeometry,
} from '../../geometry/index.js';
import type { FeatureIterator, Reader, ReaderInputs } from '../index.js';

/** A Shapefile Header describing the internal data */
export interface SHPHeader {
  length: number;
  version: number;
  shpCode: number;
  bbox: BBox3D;
}

/** A Shapefile Row explaining how to read the feature */
export interface SHPRow {
  id: number;
  len: number;
  type: number;
  data: DataView;
}

/** M-Value store */
export interface SHPMValue extends MValue {
  value: number;
}

/**
 * # The Shapefile Reader
 *
 * ## Description
 * Reads data from a shapefile implementing the {@link FeatureIterator} interface
 *
 * ## Usage
 *
 * NOTE: It's recommended to not parse the shapefile directly but instead:
 * - `import { shapefileFromURL } from 'gis-tools-ts';`
 * - `import { shapefileFromPath } from 'gis-tools-ts/file';`
 *
 * This ensures the other files paired with the shapefile are loaded to properly handle the
 * projection and properties data.
 *
 * ## Direct Usage
 *
 * ```ts
 * import { ShapeFileReader, DataBaseFile, Transformer } from 'gis-tools-ts';
 * import { FileReader } from 'gis-tools-ts/file';
 * // or use the MMapReader if using Bun:
 * // import { MMapReader } from 'gis-tools-ts/mmap';
 *
 * const transform = new Transformer();
 * const dbf = new DataBaseFile(new FileReader('./data.dbf'), 'utf-8');
 * const reader = new ShapeFileReader(new FileReader('./data.shp'), dbf, transform);
 *
 * // read all the features
 * for await (const feature of reader) {
 *   console.log(feature);
 * }
 * ```
 *
 * ## Links
 * - https://en.wikipedia.org/wiki/Shapefile
 */
export class ShapeFileReader<
  M = Record<string, unknown>,
  P extends Properties = Properties,
> implements FeatureIterator<M, SHPMValue, P> {
  reader: Reader;
  #header!: SHPHeader;
  rows: number[] = [];
  /**
   * @param input - the input data structure to parse
   * @param dbf - the dbf file
   * @param transform - transform mechanics if they exist
   */
  constructor(
    input: ReaderInputs,
    public dbf?: DataBaseFile,
    public transform?: Transformer,
  ) {
    this.reader = toReader(input);
    this.#parseHeader();
    this.#getRows();
  }

  /**
   * Return a shallow copy of the header data
   * @returns - a shallow copy of the header data
   */
  getHeader(): SHPHeader {
    return { ...this.#header };
  }

  /**
   * Return all the features in the shapefile
   * @returns - a collection of VectorFeatures
   */
  async getFeatureCollection(): Promise<VectorFeatureCollection<M, SHPMValue, P>> {
    const featureCollection: VectorFeatureCollection<M, SHPMValue, P> = {
      type: 'FeatureCollection',
      features: [],
      bbox: this.#header.bbox,
    };

    for await (const feature of this) featureCollection.features.push(feature);

    return featureCollection;
  }

  /**
   * Iterate over all features in the shapefile
   * @yields {VectorFeature}
   */
  async *[Symbol.asyncIterator](): AsyncGenerator<VectorFeature<M, SHPMValue, P>> {
    for (let i = 0; i < this.rows.length; i++) {
      const feature = this.#parseRow(this.rows[i], i);
      if (feature !== undefined) yield feature;
    }
  }

  /** Internal parse for the header */
  #parseHeader(): void {
    const { reader } = this;
    this.#header = {
      length: reader.getInt32(6 << 2) << 1,
      version: reader.getInt32(7 << 2, true),
      shpCode: reader.getInt32(8 << 2, true),
      bbox: [
        reader.getFloat64(9 << 2, true),
        reader.getFloat64(11 << 2, true),
        reader.getFloat64(13 << 2, true),
        reader.getFloat64(15 << 2, true),
        reader.getFloat64(17 << 2, true),
        reader.getFloat64(19 << 2, true),
      ],
    };
    if (this.#header.shpCode > 20) {
      this.#header.shpCode -= 20;
    }
  }

  /** Internal parser to build all the row offsets */
  #getRows(): void {
    const { reader, rows } = this;
    let offset = 100;
    const len = reader.byteLength - 8;
    while (offset <= len) {
      const offsetLength = reader.getInt32(offset + 4) << 1;
      const type = reader.getInt32(offset + 8, true);
      if (offsetLength === 0) break;
      if (type !== 0) rows.push(offset);
      offset += 8 + offsetLength;
    }
  }

  /**
   * Get a row
   * @param offset - offset of the row
   * @returns - the row if it exists
   */
  #getRow(offset: number): undefined | SHPRow {
    const { reader } = this;
    const id = reader.getInt32(offset);
    const len = reader.getInt32(offset + 4) << 1;
    if (len === 0 || offset + len + 8 > reader.byteLength) return;
    return {
      id,
      len,
      data: reader.slice(offset + 12, offset + 12 + len - 4),
      type: reader.getInt32(offset + 8, true),
    };
  }

  /**
   * Parse a row
   * @param rowOffset - the row to get and parse
   * @param index - the index of the feature
   * @returns - the parsed feature
   */
  #parseRow(rowOffset: number, index: number): VectorFeature<M, SHPMValue, P> | undefined {
    const row = this.#getRow(rowOffset);
    if (row === undefined) return;
    const { id, type, data } = row;
    const geometry = this.#parseGeometry(type, data);
    if (geometry === undefined) return;

    return {
      id,
      type: 'VectorFeature',
      properties: (this.dbf?.getProperties(index) ?? {}) as P,
      geometry,
    };
  }

  /**
   * Parse a shape geometry
   * @param type - the shape type
   * @param data - the shape data to parse
   * @returns - the parsed geometry if its valid
   */
  #parseGeometry(type: number, data: DataView): undefined | VectorGeometry<SHPMValue> {
    // ESRI Type flags:
    const is3D = type >= 11 && type <= 18;
    const hasM = (type >= 21 && type <= 28) || is3D;

    if (type === 1 || type === 11 || type === 21) {
      const mOffset = type === 21 ? 16 : is3D && data.byteLength >= 32 ? 24 : undefined;
      const coordinates = this.#parsePoint(data, 0, is3D ? 16 : undefined, mOffset);
      return { type: 'Point', is3D, coordinates };
    } else if (type === 8 || type === 18 || type === 28) {
      return this.#parseMultiPoint(data, is3D, hasM);
    } else if (
      type === 3 ||
      type === 5 ||
      type === 13 ||
      type === 15 ||
      type === 23 ||
      type === 25
    ) {
      const isPoly = type === 5 || type === 15 || type === 25;
      return this.#parseMultiLine(data, isPoly, is3D, hasM);
    } else throw new Error('invalid shape type');
  }

  /**
   * Parse a point
   * @param data - the raw data to decode
   * @param offset - the offset of the point to decode
   * @param offset3D - if provided, the offset of the Z value
   * @param offsetM - if provided, the offset of the M value
   * @returns - the decoded point
   */
  #parsePoint(
    data: DataView,
    offset: number,
    offset3D?: number,
    offsetM?: number,
  ): VectorPoint<SHPMValue> {
    const m = offsetM !== undefined ? data.getFloat64(offsetM, true) : undefined;
    const point: VectorPoint<SHPMValue> = {
      x: data.getFloat64(offset, true),
      y: data.getFloat64(offset + 8, true),
      z: offset3D !== undefined ? data.getFloat64(offset3D, true) : undefined,
      m: m !== undefined && m !== -Number.MAX_VALUE ? { value: m } : undefined,
    };
    return this.transform?.forward(point) ?? point;
  }

  /**
   * Parse a multi-point
   * @param data - the raw data to decode
   * @param is3D - is the shape a 3D shape
   * @param hasM - does the shape contain M data
   * @returns - the decoded point or multi-point
   */
  #parseMultiPoint(
    data: DataView,
    is3D = false,
    hasM = false,
  ): undefined | VectorPointGeometry<SHPMValue> | VectorMultiPointGeometry<SHPMValue> {
    const numPoints = data.getInt32(32, true);
    if (numPoints === 0) return;
    let offset = 36;
    let zOffset = 36 + 16 * numPoints;

    // Grab the min-max bounds
    const mins = this.#parsePoint(data, 0);
    const maxs = this.#parsePoint(data, 16);
    let bbox: BBOX = [mins.x, mins.y, maxs.x, maxs.y, 0, 0];

    if (is3D) {
      bbox[4] = data.getFloat64(zOffset, true);
      bbox[5] = data.getFloat64(zOffset + 8, true);
      zOffset += 16;
    }

    // M array block starts cleanly past the XY array and any Z Range/Z Array blocks
    let mOffset = 36 + 16 * numPoints + (is3D ? 16 + 8 * numPoints : 0);
    // Safe guard to check if the optional M block actually exists in this record buffer
    const holdsM = hasM && mOffset + 16 + 8 * numPoints <= data.byteLength;
    if (holdsM) mOffset += 16; // Skip past the 16-byte M Range [minM, maxM] block

    const coordinates: VectorMultiPoint<SHPMValue> = [];
    let index = 0;
    while (index < numPoints) {
      const point = this.#parsePoint(
        data,
        offset,
        is3D ? zOffset : undefined,
        holdsM ? mOffset : undefined,
      );
      coordinates.push(point);
      offset += 16;
      if (is3D) {
        zOffset += 8;
        bbox = extendBBox(bbox, point);
      }
      if (holdsM) mOffset += 8;
      index++;
    }

    if (numPoints === 1) {
      return { type: 'Point', is3D, coordinates: coordinates[0], bbox };
    } else {
      return { type: 'MultiPoint', is3D, coordinates, bbox };
    }
  }

  /**
   * Parse a multi-line
   * @param data - the raw data to decode
   * @param isPoly - is the shape a polygon or line(s)
   * @param is3D - is the shape a 3D shape
   * @param hasM - does the shape contain M data
   * @returns - the decoded point or multi-point
   */
  #parseMultiLine(
    data: DataView,
    isPoly: boolean,
    is3D: boolean,
    hasM = false,
  ):
    | undefined
    | VectorLineStringGeometry<SHPMValue>
    | VectorMultiLineStringGeometry<SHPMValue>
    | VectorPolygonGeometry<SHPMValue> {
    const numParts = data.getInt32(32, true); // The number of rings/parts
    const numPoints = data.getInt32(36, true); // The total number of points
    if (numPoints === 0 || numParts === 0) return;

    // Restore original working offsets
    let offset = 40 + 4 * numParts; // Points array starts exactly after the parts directory
    let zOffset = offset + 16 * numPoints; // Z block starts after all XY points

    // Grab the min-max bounds
    const mins = this.#parsePoint(data, 0);
    const maxs = this.#parsePoint(data, 16);
    let bbox: BBOX = [mins.x, mins.y, maxs.x, maxs.y, 0, 0];

    if (is3D) {
      bbox[4] = data.getFloat64(zOffset, true);
      bbox[5] = data.getFloat64(zOffset + 8, true);
      zOffset += 16; // Skip past the 16-byte Z Range block
    }

    // Compute the exact trailing M block offset
    let mOffset = 40 + 4 * numParts + 16 * numPoints + (is3D ? 16 + 8 * numPoints : 0);
    // Safe guard to verify the optional M block actually exists in this record buffer
    const holdsM = hasM && mOffset + 16 + 8 * numPoints <= data.byteLength;
    if (holdsM) mOffset += 16; // Skip past the 16-byte M Range [minM, maxM] block

    // Parse the parts directory correctly starting at byte 40
    const parts: number[] = [];
    let done = 0;
    while (done < numParts) {
      parts.push(data.getInt32(40 + done * 4, true));
      done++;
    }

    // Build coordinates
    let index = 0;
    const coordinates: VectorMultiLineString<SHPMValue> = [];
    for (let i = 0; i < numParts; i++) {
      const partEnd = parts[i + 1] ?? numPoints;
      const line: VectorLineString<SHPMValue> = [];

      while (index < partEnd) {
        const point = this.#parsePoint(
          data,
          offset,
          is3D ? zOffset : undefined,
          holdsM ? mOffset : undefined,
        );
        line.push(point);
        offset += 16;
        if (is3D) {
          zOffset += 8;
          bbox = extendBBox(bbox, point);
        }
        if (holdsM) mOffset += 8;
        index++;
      }
      coordinates.push(line);
    }

    if (!isPoly && numParts === 1) {
      return { type: 'LineString', is3D, coordinates: coordinates[0], bbox };
    } else {
      return { type: isPoly ? 'Polygon' : 'MultiLineString', is3D, coordinates, bbox };
    }
  }
}

File: rust/readers/shapefile/shp.rs
use super::DataBaseFile;
use crate::{
    parsers::{BufferReader, FeatureReader, Reader},
    proj::Transformer,
};
use alloc::{vec, vec::Vec};
use core::marker::PhantomData;
use s2json::{
    BBox3D, Face, MValue, MValueCompatible, Properties, VectorFeature, VectorFeatureType,
    VectorGeometry, VectorGeometryType, VectorLineString, VectorMultiLineString, VectorMultiPoint,
    VectorMultiPointGeometry, VectorPoint, VectorPointGeometry,
};

/// A Shapefile Header describing the internal data
#[derive(Debug, Clone, PartialEq)]
pub struct SHPHeader {
    /// The length of the file
    pub length: u64,
    /// The shapefile version
    pub version: i32,
    /// The shape code
    pub shp_code: i32,
    /// The bounding box
    pub bbox: BBox3D,
}

/// A Shapefile Row explaining how to read the feature
#[derive(Debug)]
pub struct SHPRow {
    id: u64,
    #[allow(dead_code)]
    len: u64,
    _type: i32,
    data: Vec<u8>,
}

/// # The Shapefile Reader
///
/// ## Description
/// Reads data from a shapefile
///
/// Implements the [`FeatureReader`] trait
///
/// ## Usage
///
/// NOTE: It's recommended to not parse the shapefile directly but instead:
/// - [`crate::readers::file::shapefile_from_path`]
/// - [`crate::readers::mmap::shapefile_from_path`]
///
/// This ensures the other files paired with the shapefile are loaded to properly handle the
/// projection and properties data.
///
/// ## Usage
///
/// The methods you have access to:
/// - [`ShapeFileReader::new`]: Create a new ShapeFileReader
/// - [`ShapeFileReader::get_header`]: Get the file header data
/// - [`ShapeFileReader::iter`]: Iterate over the features in the shapefile
///
/// ### From Path (Recommended as it will ensure to pull in associated files):
/// ```rust
/// use gistools::{parsers::{FileReader, FeatureReader}, readers::{ShapeFileReader, file::shapefile_from_path}};
/// use s2json::MValue;
/// use std::{collections::BTreeMap, path::PathBuf};
///
/// let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
/// path.push("tests/readers/shapefile/fixtures/utf.shp");
/// let path_str = path.to_str().unwrap();
///
/// #[derive(Default, Debug, Clone, MValue, PartialEq)]
/// struct Props {
///     field: String,
/// }
///
/// let shp: ShapeFileReader<FileReader, Props> =
///     shapefile_from_path(path_str, BTreeMap::from([("a".into(), "b".into())]));
///
/// let features: Vec<_> = shp.iter().collect();
/// assert_eq!(features.len(), 2);
/// ```
///
/// ## Links
/// - <https://en.wikipedia.org/wiki/Shapefile>
#[derive(Debug, Clone)]
pub struct ShapeFileReader<T: Reader, P: MValueCompatible = Properties> {
    /// The input reader
    reader: T,
    header: SHPHeader,
    dbf: Option<DataBaseFile<T, P>>, // Use the same lifetime for dbf
    transform: Option<Transformer>,
    row_offsets: Vec<u64>,
    _phantom: PhantomData<VectorFeature<(), P, MValue>>,
}
impl<T: Reader, P: MValueCompatible> ShapeFileReader<T, P> {
    /// Create a new Shapefile Reader
    pub fn new(
        mut reader: T,
        dbf: Option<DataBaseFile<T, P>>,
        transform: Option<Transformer>,
    ) -> ShapeFileReader<T, P> {
        let header = ShapeFileReader::<T, P>::parse_header(&mut reader);
        let row_offsets = ShapeFileReader::<T, P>::parse_row_offsets(&mut reader);
        ShapeFileReader::<T, P> {
            reader,
            header,
            dbf,
            row_offsets,
            transform,
            _phantom: PhantomData,
        }
    }

    /// Return a reference to the header
    pub fn get_header(&self) -> &SHPHeader {
        &self.header
    }

    /// Returns a header object from the reader
    fn parse_header(reader: &mut T) -> SHPHeader {
        let mut header = SHPHeader {
            length: (reader.int32_be(Some(6 << 2)) << 1) as u64,
            version: reader.int32_le(Some(7 << 2)),
            shp_code: reader.int32_le(Some(8 << 2)),
            bbox: BBox3D::new(
                reader.f64_le(Some(9 << 2)),
                reader.f64_le(Some(11 << 2)),
                reader.f64_le(Some(13 << 2)),
                reader.f64_le(Some(15 << 2)),
                reader.f64_le(Some(17 << 2)),
                reader.f64_le(Some(19 << 2)),
            ),
        };
        if header.shp_code > 20 {
            header.shp_code -= 20;
        }

        header
    }

    /// Returns the rows starting positions from the reader
    fn parse_row_offsets(reader: &mut T) -> Vec<u64> {
        let mut res = vec![];

        let mut offset = 100;
        let len = reader.len() - 8;
        while offset <= len {
            let offset_length = (reader.int32_be(Some(offset + 4)) << 1) as u64;
            let _type = reader.int32_le(Some(offset + 8));
            if offset_length == 0 {
                break;
            }
            if _type != 0 {
                res.push(offset);
            }
            offset += 8 + offset_length;
        }

        res
    }

    fn parse_row(&self, index: usize) -> Option<VectorFeature<(), P, MValue>> {
        if index >= self.row_offsets.len() {
            return None;
        }
        let row_offset = self.row_offsets.get(index)?;
        let SHPRow { id, _type, data, .. } = self.get_row(*row_offset)?;
        let geometry = self.parse_geometry(_type, &data);
        geometry.as_ref()?;
        let mut properties: P = P::default();
        if let Some(dbf) = &self.dbf
            && let Some(props) = dbf.get_properties(index as u64)
        {
            properties = props;
        }

        Some(VectorFeature {
            id: Some(id),
            _type: VectorFeatureType::VectorFeature,
            face: Face::WM,
            properties,
            geometry: geometry.unwrap(),
            metadata: None,
        })
    }

    /// Get a row
    fn get_row(&self, offset: u64) -> Option<SHPRow> {
        let id = self.reader.int32_be(Some(offset)) as u64;
        let len = (self.reader.int32_be(Some(offset + 4)) << 1) as u64;
        if len == 0 || offset + len + 8 > self.reader.len() {
            return None;
        }
        Some(SHPRow {
            id,
            len,
            data: self.reader.slice(Some(offset + 12), Some(offset + 12 + len - 4)),
            _type: self.reader.int32_le(Some(offset + 8)),
        })
    }

    fn parse_geometry(&self, _type: i32, data: &[u8]) -> Option<VectorGeometry<MValue>> {
        let reader: BufferReader = data.into();

        // ESRI Type flags:
        let is_3d = (11..=18).contains(&_type);
        let has_m = (21..=28).contains(&_type) || is_3d;

        if _type == 1 || _type == 11 || _type == 21 {
            let m_offset = if _type == 21 {
                Some(16)
            } else if is_3d && data.len() >= 32 {
                Some(24)
            } else {
                None
            };

            Some(VectorGeometry::Point(VectorPointGeometry {
                _type: VectorGeometryType::Point,
                is_3d,
                coordinates: self.parse_point(
                    &reader,
                    0,
                    if is_3d { Some(16) } else { None },
                    m_offset,
                ),
                ..Default::default()
            }))
        } else if _type == 8 || _type == 18 || _type == 28 {
            self.parse_multi_point(&reader, is_3d, has_m)
        } else if _type == 3
            || _type == 5
            || _type == 13
            || _type == 15
            || _type == 23
            || _type == 25
        {
            let is_poly = _type == 5 || _type == 15 || _type == 25;
            self.parse_multi_line(&reader, is_poly, is_3d, has_m)
        } else {
            panic!("invalid shape type: {}", _type);
        }
    }

    /// Parse a point
    fn parse_point(
        &self,
        data: &BufferReader,
        offset: u64,
        offset_3d: Option<u64>,
        offset_m: Option<u64>,
    ) -> VectorPoint<MValue> {
        let z = offset_3d.map(|off| data.f64_le(Some(off)));

        let m_value = offset_m.map(|off| data.f64_le(Some(off)));
        let m = match m_value {
            Some(m_val) if m_val != f64::MIN => {
                Some(MValue::from([("value".into(), m_val.into())]))
            }
            _ => None,
        };

        let mut point =
            VectorPoint::new(data.f64_le(Some(offset)), data.f64_le(Some(offset + 8)), z, m);

        if let Some(transformer) = &self.transform {
            point = transformer.forward(&point);
        }

        point
    }

    fn parse_multi_point(
        &self,
        data: &BufferReader,
        is_3d: bool,
        has_m: bool,
    ) -> Option<VectorGeometry<MValue>> {
        let num_points = data.int32_le(Some(32)) as u64;
        if num_points == 0 {
            return None;
        }

        let mut offset = 36;
        let mut z_offset = 36 + 16 * num_points;

        // Grab the min-max bounds
        let mins = self.parse_point(data, 0, None, None);
        let maxs = self.parse_point(data, 16, None, None);
        let mut bbox = BBox3D::new(mins.x, mins.y, maxs.x, maxs.y, 0., 0.);

        if is_3d {
            bbox.near = data.f64_le(Some(z_offset));
            bbox.far = data.f64_le(Some(z_offset + 8));
            z_offset += 16;
        }

        // M array block starts cleanly past the XY array and any Z Range/Z Array blocks
        let mut m_offset = 36 + 16 * num_points + if is_3d { 16 + 8 * num_points } else { 0 };

        // Safe guard to check if the optional M block actually exists in this record buffer
        // Assuming data has a .len() method returning total byte length
        let holds_m = has_m && m_offset + 16 + 8 * num_points <= data.len();
        if holds_m {
            m_offset += 16; // Skip past the 16-byte M Range [minM, maxM] block
        }

        let mut coordinates: VectorMultiPoint<MValue> = vec![];
        let mut index = 0;
        while index < num_points {
            let point = self.parse_point(
                data,
                offset,
                if is_3d { Some(z_offset) } else { None },
                if holds_m { Some(m_offset) } else { None },
            );

            offset += 16;
            if is_3d {
                z_offset += 8;
                bbox.extend_from_point(&point);
            }
            if holds_m {
                m_offset += 8;
            }
            coordinates.push(point);
            index += 1;
        }

        if num_points == 1 {
            Some(VectorGeometry::Point(VectorPointGeometry {
                _type: VectorGeometryType::Point,
                is_3d,
                coordinates: core::mem::take(&mut coordinates[0]),
                bbox: Some(bbox),
                ..Default::default()
            }))
        } else {
            Some(VectorGeometry::MultiPoint(VectorMultiPointGeometry {
                _type: VectorGeometryType::MultiPoint,
                is_3d,
                coordinates,
                bbox: Some(bbox),
                ..Default::default()
            }))
        }
    }

    fn parse_multi_line(
        &self,
        data: &BufferReader,
        is_poly: bool,
        is_3d: bool,
        has_m: bool,
    ) -> Option<VectorGeometry<MValue>> {
        let num_parts = data.int32_le(Some(32)) as u64; // The number of rings/parts
        let num_points = data.int32_le(Some(36)) as u64; // The total number of points
        if num_points == 0 || num_parts == 0 {
            return None;
        }

        // Restore original working offsets
        let mut offset = 40 + 4 * num_parts; // Points array starts exactly after the parts directory
        let mut z_offset = offset + 16 * num_points; // Z block starts after all XY points

        // Grab the min-max bounds
        let mins = self.parse_point(data, 0, None, None);
        let maxs = self.parse_point(data, 16, None, None);
        let mut bbox = BBox3D::new(mins.x, mins.y, maxs.x, maxs.y, 0., 0.);

        if is_3d {
            bbox.near = data.f64_le(Some(z_offset));
            bbox.far = data.f64_le(Some(z_offset + 8));
            z_offset += 16; // Skip past the 16-byte Z Range block
        }

        // Compute the exact trailing M block offset
        let mut m_offset =
            40 + 4 * num_parts + 16 * num_points + if is_3d { 16 + 8 * num_points } else { 0 };

        // Safe guard to verify the optional M block actually exists in this record buffer
        let holds_m = has_m && m_offset + 16 + 8 * num_points <= data.len();
        if holds_m {
            m_offset += 16; // Skip past the 16-byte M Range [minM, maxM] block
        }

        // Parse the parts directory correctly starting at byte 40
        let mut parts: Vec<u64> = vec![];
        let mut done = 0;
        while done < num_parts {
            parts.push(data.int32_le(Some(40 + done * 4)) as u64);
            done += 1;
        }

        // Build coordinates
        let mut index = 0;
        let mut coordinates: VectorMultiLineString<MValue> = vec![];
        for i in 0..num_parts {
            let part_end = parts.get(i as usize + 1).unwrap_or(&num_points);
            let mut line: VectorLineString<MValue> = vec![];

            while index < *part_end {
                let point = self.parse_point(
                    data,
                    offset,
                    if is_3d { Some(z_offset) } else { None },
                    if holds_m { Some(m_offset) } else { None },
                );

                offset += 16;
                if is_3d {
                    z_offset += 8;
                    bbox.extend_from_point(&point);
                }
                if holds_m {
                    m_offset += 8;
                }
                line.push(point);
                index += 1;
            }
            coordinates.push(line);
        }

        if !is_poly && num_parts == 1 {
            Some(VectorGeometry::new_linestring(core::mem::take(&mut coordinates[0]), Some(bbox)))
        } else if is_poly {
            Some(VectorGeometry::new_polygon(coordinates, Some(bbox)))
        } else {
            Some(VectorGeometry::new_multilinestring(coordinates, Some(bbox)))
        }
    }
}
/// The GPX Iterator tool
#[derive(Debug)]
pub struct ShapefileIterator<'a, T: Reader, P: MValueCompatible> {
    reader: &'a ShapeFileReader<T, P>,
    index: usize,
    stride: usize,
}
impl<T: Reader, P: MValueCompatible> Iterator for ShapefileIterator<'_, T, P> {
    type Item = VectorFeature<(), P, MValue>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(feature) = self.reader.parse_row(self.index) {
            self.index += self.stride;
            return Some(feature);
        }
        None
    }
}
/// A feature reader trait with a callback-based approach
impl<T: Reader, P: MValueCompatible> FeatureReader<(), P, MValue> for ShapeFileReader<T, P> {
    type FeatureIterator<'a>
        = ShapefileIterator<'a, T, P>
    where
        T: 'a,
        P: 'a;

    fn iter(&self) -> Self::FeatureIterator<'_> {
        ShapefileIterator { reader: self, index: 0, stride: 1 }
    }
    // The assumption here is that the reader has been cloned already
    fn par_iter(&self, pool_size: usize, thread_id: usize) -> Self::FeatureIterator<'_> {
        ShapefileIterator { reader: self, index: thread_id, stride: pool_size }
    }
}
